File: Ruby/ClusterSignature.rb
class ClusterSignature

  def self.generate_signatures (log_entry)
    signatures = []
    log_entry.words.each_index do |i|
      static_words = log_entry.words.clone
      static_words.delete_at(i)
      signatures << ClusterSignature.new(i, static_words)
    end
    signatures
  end

  def initialize (varword_index, static_words)
    @varword_index = varword_index
    @static_words = static_words
  end

  ## Enable use as Hash key
  attr_reader :varword_index, :static_words
  def == (other)
    self.class === other and
      other.varword_index == @varword_index and
      other.static_words == @static_words
  end
  alias eql? ==

  def hash
    @varword_index.hash ^ @static_words.hash
  end
end

File: Node-js/ClusterSignature.js
class ClusterSignature {
  constructor(varword_index, static_words) {
    this.varword_index = varword_index
    this.static_words = static_words
  }

  static generate_signatures(log_entry) {
    let signatures = []
    log_entry.words.forEach((e,i,a) => {
      let static_words = [].concat(log_entry.words)
      static_words.splice(i, 1)
      signatures.push(new ClusterSignature(i, static_words))
    })
    return signatures
  }

  toString() {
    return `ClusterSignature: varword_index=${this.varword_index} static_words=${this.static_words}`
  }
}

module.exports = ClusterSignature
